ballerina/utils.bal

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2025 WSO2 LLC (http://www.wso2.com).
+//
+// WSO2 LLC. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/ai;
+
+# Converts metadata filters to Weaviate compatible filter format
+#
+# + filters - The metadata filters containing filter conditions and logical operators
+# + return - A map representing the converted filter structure or an error if conversion fails
+isolated function convertWeaviateFilters(ai:MetadataFilters filters) returns map<anydata>|ai:Error {
+    (ai:MetadataFilters|ai:MetadataFilter)[]? rawFilters = filters.filters;
+    if rawFilters == () || rawFilters.length() == 0 {
+        return {};
+    }
+    map<anydata>[] filterList = [];
+    foreach (ai:MetadataFilters|ai:MetadataFilter) filter in rawFilters {
+        if filter is ai:MetadataFilter {
+            map<anydata> filterMap = {};
+            string weaviateOp = check mapWeaviateOperator(filter.operator);
+            filterMap["path"] = [filter.key];
+            filterMap["operator"] = weaviateOp;
+            filterMap["valueText"] = filter.value;
+            filterList.push(filterMap);
+            continue;
+        }
+        map<anydata> nestedFilter = check convertWeaviateFilters(filter);
+        if nestedFilter.length() > 0 {
+            filterList.push(nestedFilter);
+        }
+    }
+    if filterList.length() == 0 {
+        return {};
+    }
+    if filterList.length() == 1 {
+        return filterList[0];
+    }
+    string weaviateCondition = check mapWeaviateCondition(filters.condition);
+    map<anydata> result = {
+        "operator": weaviateCondition, 
+        "operands": filterList
+    };
+    return result;
+}
+
+# Maps metadata filter operators to Weaviate compatible operators
+#
+# + operation - The metadata filter operator to map
+# + return - The Weaviate compatible operator or an error if the operator is not supported
+isolated function mapWeaviateOperator(string operation) returns string|ai:Error {
+    match operation {
+        "==" => {
+            return "Equal";
+        }
+        "!=" => {
+            return "NotEqual";
+        }
+        ">" => {
+            return "GreaterThan";
+        }
+        ">=" => {
+            return "GreaterThanEqual";
+        }
+        "<" => {
+            return "LessThan";
+        }
+        "<=" => {
+            return "LessThanEqual";
+        }
+        _ => {

```

Comment:
I prefer to have this after the matchcase

Reply:
Fixed in [e6028fb](https://github.com/ballerina-platform/module-ballerinax-ai.weaviate/pull/1/commits/e6028fb6b80cc729f2eac69bf67a6fd3c5ffbcff)